pkg/rancher/migrations.go

```diff
@@ -0,0 +1,79 @@
+package rancher
+
+import (
+	"github.com/mcuadros/go-version"
+	"github.com/rancher/rancher/pkg/auth/tokens"
+	v3 "github.com/rancher/rancher/pkg/generated/controllers/management.cattle.io/v3"
+	rancherversion "github.com/rancher/rancher/pkg/version"
+	controllerv1 "github.com/rancher/wrangler/pkg/generated/controllers/core/v1"
+	"github.com/sirupsen/logrus"
+	v1 "k8s.io/api/core/v1"
+	k8serror "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+)
+
+const (
+	cattleNamespace        = "cattle-system"
+	upgradeMigrationConfig = "migrated"
+	forceLogoutKey = "forcedLogout"
+)
+
+// forceUpgradeLogout will delete all dashboard tokens forcing a logout.  This is useful when there is a major frontend
+//upgrade and we want all users to be sent to a central point.  This function will check for the `upgradeMigrationConfig`
+// configuration map and only run if the last migrated version is lower than the given `migrationVersion`.
+func forceUpgradeLogout(configMapController controllerv1.ConfigMapController, tokenController v3.TokenController, migrationVersion string) (err error) {
+	var upgradeMigrationConfigMap *v1.ConfigMap
+
+	upgradeMigrationConfigMap, err = configMapController.Get(cattleNamespace, upgradeMigrationConfig, metav1.GetOptions{})
+	if err != nil {
+		if k8serror.IsNotFound(err) { // if the config map does not exist we need to initialize it
+			upgradeMigrationConfigMap.Data = map[string]string{}
+		} else {
+			return
+		}
+	}
+
+	// clear out any revision ids or other metadata we don't want to save or initialize the cm for the first time
+	upgradeMigrationConfigMap = &v1.ConfigMap{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: upgradeMigrationConfig,
+			Namespace: cattleNamespace,
+		},
+		Data: upgradeMigrationConfigMap.Data,
+	}
+
+	// we only want to migrate if the previous installation was earlier than the migration version
+	if lastMigration, ok := upgradeMigrationConfigMap.Data[forceLogoutKey]; ok {
+		if lastMigration == "dev" || version.Compare(migrationVersion, lastMigration, "<=") {
+			return
+		}
+	}
```

Comment:
This section is awkward, get the config map, check the value and exit. Removing the data defeats the purpose of getting the existing version and also doesn't allow k8s to block rancher from overwriting the same change. 
If the configmap doesn't exist then create the object with the value as expected. If it does exist then update the value and save. 
This is also moving to a cache so a deepcopy will be required
There is no update call so if it exists the create will always fail.